### ISSUE TITLE
feat(tutorials): incorporate LightGBM and Random Forest companion baselines and ensembling in Colab

### ISSUE THREAD

**Issue by kurtvalcorza** [kurtvalcorza/tabicl-classifier-pipeline#13]
### Scope & Context
**Scope: Google Colab tutorial workflow (`tutorials/tabiclv2_classifier_colab.ipynb`) and `tutorials/README.md` only.**  
- The headless container contract, backend engine, and downstream inference tutorial (`tabiclv2_classifier_artifact_inference_colab.ipynb`) remain **completely untouched**.
- The artifact format identifier `ARTIFACT_FORMAT = 'tabicl-dimer-classifier-v1'` and exported artifact structure (`tabicl-classifier-artifact.zip`) are **strictly preserved** as standard TabICL context payloads without embedded tree boosters. Dual-mode artifact packaging is deferred to a separate issue.

### Feature Description
Incorporate lightweight classical tree models (LightGBM and Random Forest) directly into the standalone TabICLv2 classifier Colab tutorial notebook as an additive benchmark leaderboard and in-memory post-hoc ensembling demonstration alongside the TabICLv2 foundation model.

### Key Capabilities in the Colab Tutorial
1. **Dependency Pinning:** Explicitly add `"lightgbm>=4.0,<4.8"` as a separate quoted token in the existing `%pip install` cell, strictly preserving the CI-required `'tabicl[finetune]==2.1.1'` literal.
2. **Benchmark Leaderboard Cell:** 
   - Fit LightGBM and Random Forest on `train_encoded` (the exact support context used by TabICLv2).
   - Evaluate on `holdout_encoded` (and independent `test_encoded` when present).
   - Display a side-by-side leaderboard (Accuracy, ROC-AUC) alongside the holdout one-row resolution ($1/N$, e.g. 0.88% on 114 rows).
   - Record reproducible latency with runtime device (GPU/CPU) and row/batch counts printed per row.
3. **Label & Probability Alignment:**
   - Explicitly reorder LightGBM predicted probability columns to match the exact TabICLv2 class ordering (`model.classes_`) before blending, and assert class order agreement to eliminate silent multiclass permutation bugs.
4. **In-Memory Post-Hoc Ensembling Cell:** 
   - Optimize convex blend weights on `holdout_encoded` against a declared objective (Accuracy or ROC-AUC).
   - Measure and report the actual empirical delta against standalone models alongside the one-row resolution.
   - When independent `test_encoded` is available, test whether holdout gains generalize; without an independent test partition, explicitly label the blended score as selection-biased demonstration evidence.
   - **State Isolation:** The blend is evaluated purely in-memory and does not alter the exported `tabicl-classifier-artifact.zip`.
5. **In-Memory Latency Comparison:**
   - Demonstrate in-memory evaluation speed comparing foundation model evaluation vs. ultra-fast CPU tree inference with the resolved device printed per row.
6. **Documentation:** Update `tutorials/README.md` to reflect the companion baseline, probability alignment, and ensembling tutorial steps.

**Comment by kurtvalcorza** [comment]
### RFC re-review — **APPROVE WITH MINOR CHANGES**

The tutorial-only scope resolves my previous artifact/serving-contract concerns. The current notebook already has a clean train → holdout → optional independent-test evaluation structure, so the baseline/ensemble demo can fit naturally into it.

Three adjustments before implementation:

1. **Replace “out-of-fold accuracy improvements.”** This notebook does not currently perform cross-validation/OOF evaluation. TabICL registers `train_encoded` as context and evaluates on `holdout_encoded`, with `test_encoded` reserved as an independent report when present. Call this a holdout/post-hoc ensemble demonstration unless the RFC explicitly adds K-fold OOF machinery.
2. **Align probability columns before blending.** For binary/multiclass classification, LightGBM probability columns must be reordered to the exact TabICL class order (`model.classes_`) before convex blending; otherwise the ensemble can silently mix probabilities for different labels.
3. **Measure improvement; do not assume it.** Select the blend weight on the holdout, report the actual delta (including zero/negative), and use the independent test—when present—to show whether the holdout-selected blend generalizes. Without a test split, label the ensemble score as selection-biased/demo evidence.

Also add/pin `lightgbm` in the Colab install cell; Random Forest is already covered by scikit-learn.

With those changes, this RFC is implementation-ready.

**Comment by kurtvalcorza** [comment]
Thank you for the review. The RFC specification has been updated in the issue description above to incorporate all points:

1. **Evaluation Splitting Clarification:** Removed all out-of-fold (OOF) terminology. The workflow fits LightGBM and Random Forest on `train_encoded` (the exact support context used by TabICL) and evaluates on `holdout_encoded`, with independent `test_encoded` reserved as a generalization benchmark when present.
2. **Probability Column Alignment:** Explicitly documented the requirement to reorder LightGBM predicted probability columns to match the exact TabICL class ordering (`model.classes_`) before computing convex blends, preventing label permutation errors in multiclass classification.
3. **Objective Performance Measurement:** Acceptance criteria do not assume an accuracy boost. The tutorial will measure and report the empirical performance delta (including zero or negative deltas). The blend weight will be selected on `holdout_encoded` (targeting holdout Accuracy or ROC-AUC), and generalization will be reported on `test_encoded` when present (or labeled as selection-biased demo evidence when absent).
4. **Dependency Pinning:** Explicitly added and pinned `lightgbm` in the notebook installation cell (`scikit-learn` already provides `RandomForestClassifier`).

The issue is aligned and ready for implementation.

**Comment by kurtvalcorza** [comment]
## Agent Relay — Reviewer[design] re-review after the scope revision

Re-read against the revised issue body and the live notebook, not the earlier draft. No code changed.

**The narrowing to `tutorials/` resolves most of what a backend-scoped version of this would have hit, and it puts the issue on the repo that actually owns the file.** What remains is smaller and sharper: the tutorial validators assert literal strings that three of the four capabilities would delete, and the bundled sample cannot support the "empirical boost" the RFC promises.

### What the revision resolves

- **Ownership layer.** `feat(tutorials):` on the repo that owns `tutorials/` is correct. The earlier `feat(finetuner):` framing pointed at a different repo.
- **DIMER result/report contract.** `evaluation/report.json`, `logs/run-summary.json` and `progress/epoch_*.json` are marked `stable-v1` in `dimer-runtime-contract.json`; nothing in the revised scope touches them.
- **The wizard schema.** `dimer-pipeline.json` has only `datasetPreprocessing` and `modelFinetuning` — no serving section for a `fast_cpu` flag. Out of scope now, correctly.
- **The container's time budget and model lock.** `time_limit_seconds` and the permanent base-model lock in the finetuner are untouched.

Everything below is scoped to the notebook.

### N1 — Correction to my earlier read: the notebook is *not* GPU-only

I previously said this pipeline has no CPU path. That is true of the container (`train.py` raises `TabICLv2 fine-tuning requires a CUDA GPU`), which the revision puts out of scope. In the notebook it is not: `DEVICE = "cuda" if torch.cuda.is_available() else "cpu"`, and only fine-tuning raises on CPU. A CPU-vs-ensemble latency demo is therefore feasible here. The objection to capability 3 below is about the artifact format, not feasibility.

### N2 — The artifact format id is shared with the backend and CI-pinned in both notebooks (the finding the new scope creates)

`ARTIFACT_FORMAT = 'tabicl-dimer-classifier-v1'` is a **required literal in the main notebook and in the artifact-inference notebook**, and the export cell carries the standing comment:

> `# Key/value pairs are written without spaces on purpose: the repository tests read them literally.`

The finetuner container emits that **same** format string from its own `artifact.json`. So the tutorial bundle and the backend bundle currently claim to be one format, and `tabiclv2_classifier_artifact_inference_colab.ipynb` is the consumer of both.

Capability 3 says the CPU mode uses "the embedded tree booster" — i.e. the tree ships **inside the exported bundle**. Doing that while the backend is deliberately untouched means two different payloads under one unchanged format id, and an inference notebook that must handle a bundle that may or may not contain trees with nothing in the manifest to say which.

"The backend remains untouched" is what makes this sharp rather than what removes it. If capability 3 ships, the tutorial bundle needs either its own format id or a `-v2` bump with the backend following. Silent divergence is the one option that should be off the table.

### N3 — Blend weight vs. the notebook's own selection discipline

The notebook enforces a discipline it advertises: `MIN_SELECTION_HOLDOUT_ROWS = 50`, `default:pretrained`, `holdout-too-small`, and *"The independent test split never participates in selection"* — all CI-asserted literals.

Tuning a convex blend weight on `holdout_encoded` and then reporting the blend's improvement on `holdout_encoded` is selection on the reporting set. It will show a gain by construction. The RFC needs to name the weight-fitting set and the reporting set, and define behavior when `test_encoded is None` — which is the common case for a reader bringing their own data.

### N4 — The sample is too small for the improvement to be readable

With the bundled sample the notebook prints `train=341, holdout=114, test=114`. On 114 holdout rows, one row is **0.88 %**. A convex blend that improves accuracy by "about a point" has moved one row.

Capability 2 promises a demonstrated improvement; on this data most achievable improvements are inside the resolution of the holdout. Print the one-row resolution beside the delta and let the cell say when the difference is not readable. The demonstration is still worth having — it just has to be honest about what it shows.

### N5 — LightGBM is a new dependency; the install pin is a CI-asserted literal

`%pip -q install "tabicl[finetune]==2.1.1" "pyarrow>=15" "pandas>=2" "scikit-learn>=1.4" "huggingface_hub>=0.25"` — Random Forest is already available via scikit-learn; **LightGBM is not present**.

The validator requires the literal `'tabicl[finetune]==2.1.1'` in the main notebook and `'tabicl==2.1.1'` in the inference notebook, so add LightGBM as a separate quoted token rather than editing the pinned spec:

```
%pip -q install "tabicl[finetune]==2.1.1" "lightgbm>=4.0,<4.8" "pyarrow>=15" ...
```

Keep it in the same install cell — Step 1 asserts the resolved `tabicl` version immediately afterward and a later install pass is how that assertion starts failing intermittently.

### N6 — Latency rows must carry their device

Colab hardware varies per session, and `DEVICE` is resolved at runtime. A leaderboard row showing TabICLv2 at one latency and LightGBM at another is comparing a GPU (or CPU) foundation model against a CPU tree; without the resolved device printed per row it reads as a property of the models. Cheap fix, and it makes capability 3's demo self-explanatory.

### Recommendation

**Capabilities 1 and 4 are ready to build once N1 is handled.** They are additive cells plus a README section, they fit the notebook's existing shape, and the validators do not stand in their way.

**Capability 2 needs one decision first:** where the blend weight is fitted and where it is reported, and whether the blend may ever become the exported/selected model. Answer that in the issue, not in the code review.

**Capability 3 should be split into its own issue.** It is the only one that touches the exported bundle, which is the tutorial's contract with the second notebook.

Suggested sequencing: land 1 + 4 together, open a follow-up for 2 once the split is decided, and a separate issue for 3.

**Nothing here says don't do this.** The revised scope is the right scope, and the leaderboard is the most useful part of the RFC.

---

<details>
<summary>Agent Relay pass record</summary>

```text
Agent pass: rfc-design-review-r2
Role: reviewer
Review lenses: design, spec-conformance
Mission mode: analyze
Role source: explicit
Role sequence: reviewer[design] -> integrator (decide capability 2's split) -> builder
Mission anchor: kurtvalcorza/tabicl-classifier-pipeline#13 @ updatedAt 2026-09-07T11:38:11Z (revised scope: tutorials only)
Predecessor pass: rfc-design-review (backend-scoped; superseded by the revision, not by new evidence)
Decision authority: review + record only; no mutation of source, no approval, no closure
Assurance profile: standard
Mutation boundary: allowed = this issue's comment thread. read-only = all repository source
Reviewed snapshot: tabicl-classifier-pipeline@629ea14 (tutorials/ + scripts/), tabicl-classifier-finetuner@e05a456 (format-id cross-check only)
Environment: Windows 11, gh CLI + local clones at the SHAs above; no notebook executed, no Colab run
Execution status: N/A (static review)
Claims / evidence maturity:
- INSPECTED: every finding cited to a notebook cell or a validator line in the snapshot above
- INSPECTED: validator required-literal lists read from scripts/validate_colab_tutorial.py
- ASSERTED: AutoGluon's behavior when a model's package is missing (N1)
- DISPROVED (prior pass): "this pipeline has no CPU path" — true of the container, false of the notebook (N1)
Verification contracts:
- "the new cells do not break CI" -> `python scripts/validate_colab_tutorial.py` after the change
- "the blend improves the score" -> reported on a set the weight was not fitted on, with the one-row resolution printed
Findings: 6 live, 4 resolved by the revision
Fixed: 0 | Disproved: 1 | Deferred: 0 | Blocked: 0
Verification checkpoint: no notebook was executed; every finding is source-level
Next recommended role/pass: integrator — decide capability 2's evaluation split, then builder
Provenance:
- Generated by: Claude Code
- Model: claude-opus-5
- Source snapshot: tabicl-classifier-pipeline@629ea14 (tutorials/ + scripts/), tabicl-classifier-finetuner@e05a456 (format-id cross-check only)
```

Attribution is provenance, not sign-off.
</details>

**Comment by kurtvalcorza** [comment]
## Correction — my review above was written against a superseded revision

I fetched this issue at `updatedAt 2026-09-07T11:38:11Z` and the body moved again at 11:44 while I was reading the notebook. Retargeting the findings silently would hide which snapshot each was observed against, so here is the delta instead. The review above stands as written for the 11:38 revision.

**Resolved by the current body — disregard these:**

- **N3** (blend weight vs. the notebook's selection discipline). The declared objective, the generalization check on `test_encoded`, and the explicit selection-biased labeling when no test partition exists answer it in full.
- **N5** (LightGBM dependency) — now capability 1. One implementation note survives it: `scripts/validate_colab_tutorial.py` requires the literal `'tabicl[finetune]==2.1.1'` in the main notebook and `'tabicl==2.1.1'` in the inference notebook. "Explicitly add and pin lightgbm" must therefore be a **separate quoted token** on the same `%pip` line; rewriting the pinned spec itself turns CI red.
- **N6** (latency device) — "latency with device and row/batch count" covers it.

Also worth saying plainly: the **Label & Probability Alignment** clause added to capability 3 is the right call, and it is the item I would have raised next. `LGBMClassifier.predict_proba` orders columns by its own `classes_`; blending positionally against TabICLv2's ordering is silently wrong rather than an error. Asserting the two orders match before blending is cheap insurance.

**Still live, unchanged by the revision:**

- **N2** — the artifact format id. `ARTIFACT_FORMAT = 'tabicl-dimer-classifier-v1'` is a required literal in *both* notebooks, the export cell warns that the manifest is read literally by the repository tests, and the finetuner container emits that same string. Capability 4's "embedded tree booster" puts a tree inside that bundle while the backend is deliberately untouched — two different payloads under one format id, with `tabiclv2_classifier_artifact_inference_colab.ipynb` consuming both. This needs its own format id or a `-v2` bump with the backend following.
- **N4**, reduced to a note: 114 holdout rows means one row is 0.88 %. The new "measure and report the actual delta" wording already prevents the overclaim; printing the one-row resolution beside the delta makes the number self-explanatory.
- **N1** stands as the correction it was: the notebook is not GPU-only, so capability 4 is feasible here. The objection is the format id, not the CPU mode.

Recommendation from the review above is unchanged: capabilities 1–3 and 5 are ready to build; 4 belongs in its own issue.

---

<sub>Reviewer[design] · reviewed `629ea14` (tutorials/ + scripts/) · issue body at `11:38:11Z`, corrected against `11:44:02Z` · Claude Code / claude-opus-5 · provenance, not sign-off</sub>

**Comment by kurtvalcorza** [comment]
Thank you for the thorough design review and correction note. The RFC specification in the issue description has been updated to v3, adopting your recommendations in full:

1. **Dependency Pinning (N5):** Confirmed `"lightgbm>=4.0,<4.8"` will be added as a separate quoted token on the existing `%pip install` line, leaving the CI-required `'tabicl[finetune]==2.1.1'` literal untouched.
2. **Artifact Format ID & Packaging Boundary (N2 & Recommendation):** We adopt the recommendation to **exclude artifact bundling from this issue**. The exported bundle (`tabicl-classifier-artifact.zip`) and `ARTIFACT_FORMAT = 'tabicl-dimer-classifier-v1'` remain 100% standard and untouched, preserving contract compatibility with `tabiclv2_classifier_artifact_inference_colab.ipynb` and the container backend. Dual-mode demonstration is strictly an in-memory latency comparison; artifact-level tree packaging will be tracked in a separate issue.
3. **Probability & Label Alignment:** Re-verified explicit column reordering of LightGBM probabilities to match `model.classes_` with an assertion check prior to convex blending.
4. **Resolution Reporting (N4):** The notebook will display the holdout one-row resolution ($1/N$, e.g., 0.88% on 114 rows) directly beside reported metric deltas to contextualize small score movements.
5. **Reproducible Latency Benchmarks (N6):** Benchmark tables will print the resolved execution device (GPU/CPU) per row.

The issue is updated and ready for implementation.
